apps/web/src/components/AssessmentStepTracker/utils.ts

```diff
@@ -326,3 +213,192 @@ export const sortResultsAndAddOrdinal = (
+      return result.poolSkill?.id === skillAssessment?.id;
+    });
+
+    if (assessmentResults.length) {
+      hasToAssess = true;
+      return;
+    }
+
+    assessmentResults.forEach((assessmentResult) => {
+      switch (assessmentResult.assessmentDecision) {
+        case null:
+          hasToAssess = true;
+          break;
+        case AssessmentDecision.Hold:
+          hasOnHold = true;
+          break;
+        case AssessmentDecision.Unsuccessful:
+          hasFailure = true;
+          break;
+        default:
+      }
+    });
+  });
+
+  // Could be an education requirement that is not an essential skill
+  stepResults.forEach((result) => {
+    // Any "to assess" should be marked
+    if (result.assessmentDecision === null) {
+      hasToAssess = true;
+    }
+
+    if (result.assessmentResultType === AssessmentResultType.Education) {
+      switch (result.assessmentDecision) {
+        case null:
+          hasToAssess = true;
+          break;
+        case AssessmentDecision.Hold:
+          hasOnHold = true;
+          break;
+        case AssessmentDecision.Unsuccessful:
+          hasFailure = true;
+          break;
+        default:
+      }
+    }
+  });
+
+  if (hasToAssess) {
+    return NO_DECISION;
+  }
+  if (hasOnHold) {
+    return AssessmentDecision.Hold;
+  }
+  if (hasFailure) {
+    return AssessmentDecision.Unsuccessful;
+  }
+
+  return AssessmentDecision.Successful;
+};
+
+type AssessmentMap = Map<string, CandidateAssessmentResult>;
+
+type GroupedStep = {
+  step: AssessmentStep;
+  assessments: AssessmentMap;
+};
+
+type GroupedSteps = Map<string, GroupedStep>;
+
+export const groupPoolCandidatesByStep = (pool: Pool) => {
+  const poolCandidates = pool.poolCandidates?.filter(notEmpty) ?? [];
+  const steps = pool.assessmentSteps?.filter(notEmpty) ?? [];
+
+  // Build a map for all poolCandidates on step 1
+  const allCandidatesMap: AssessmentMap = new Map();
+  poolCandidates.forEach((poolCandidate) => {
+    allCandidatesMap.set(poolCandidate.id, {
+      poolCandidate,
+      decision: NO_DECISION,
+      results: [],
+    });
+  });
+
+  // Setup the base map for steps with assessments for candidates
+  const stepMap: GroupedSteps = new Map();
+  steps.forEach((step, index) => {
+    stepMap.set(step.id, {
+      step,
+      // Step one should show all candidates regardless of results existing
+      assessments: index === 0 ? allCandidatesMap : new Map(),
+    });
+  });
+
+  poolCandidates.forEach((poolCandidate) => {
+    // Only use assessments associated with this pool
+    const assessments = poolCandidate.assessmentResults
+      ?.filter((result) =>
+        result?.assessmentStep?.id
+          ? stepMap.has(result?.assessmentStep?.id)
+          : false,
+      )
+      .filter(notEmpty);
+
+    assessments?.forEach((result) => {
+      if (!result) return;
+
+      const resultStep = result?.assessmentStep?.id
+        ? stepMap.get(result.assessmentStep.id)
+        : false;
```

Comment:
The issue here is that `id` is undefined which you cannot pass to a map. Though I guess `result?.assessmentStep?.id ?? ""` would work since we never expect an empty string as key 🤔

Reply:
Is that what typescript says? It's valid javascript in practice 😆 

```
const m = new Map();
m.set("one", 1);
  # Map(1) {'one' => 1}
m.get(undefined)
  # undefined
m.set(undefined, "test")
  # Map(2) {'one' => 1, undefined => 'test'}
m.get(undefined);
  # 'test'

Reply:
Well, I guess I could type it to allow undefined, but I was trying to be safe 😆 I think `undefined` as a key should be fine so I'll just update the type 👍 

https://github.com/GCTC-NTGC/gc-digital-talent/blob/5df5503138a93af178dbd9956b446bd111f13323/apps/web/src/components/AssessmentStepTracker/utils.ts#L305

Reply:
Updated to allow `undefined` keys to reduce the logic here, thanks for the suggestion 👍 

[allow undefined keys for grouped steps](https://github.com/GCTC-NTGC/gc-digital-talent/pull/9022/commits/3c19d968b263a0f711e6f38746558c3eba05cec5)